core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -361,14 +364,19 @@ public ReserveResult tryReserve(PendingAcquire acquire)
                         }
                         return false;
                     })
-                    .max(comparing(node -> nodesRemainingMemory.get(node.getNodeIdentifier())));
+                    .max(comparing(node -> nodesUnreservedMemory.get(node.getNodeIdentifier())));
 
             if (selectedNode.isEmpty()) {
                 return ReserveResult.NONE_MATCHING;
             }
 
             boolean selectedCandidateMatches = false;
-            if (nodesRemainingMemory.get(selectedNode.get().getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.get())) {
+
+            String selectNodeId = selectedNode.get().getNodeIdentifier();
+            boolean enoughUnreservedMemory = nodesUnreservedMemory.get(selectNodeId) >= acquire.getMemoryLease();
+            boolean enoughRuntimeMemory = !nodeMemoryPoolInfos.containsKey(selectNodeId)
+                    || nodeMemoryPoolInfos.get(selectNodeId).getFreeBytes() >= acquire.getMemoryLease();
```

Comment:
I wonder if it makes sense to leave some "head room", e.g. some percentage of the memory pool size? (say 10-15%)